presto-main/src/main/java/com/facebook/presto/execution/MemoryRevokingScheduler.java

```diff
@@ -86,33 +103,125 @@ public MemoryRevokingScheduler(
     @PostConstruct
     public void start()
     {
+        registerPeriodicCheck();
+        registerPoolListeners();
+    }
+
+    @VisibleForTesting
+    void registerPeriodicCheck()
+    {
         taskManagementExecutor.scheduleWithFixedDelay(() -> {
             try {
-                requestMemoryRevokingIfNeeded(currentTasksSupplier.get());
+                requestMemoryRevokingIfNeeded();
             }
             catch (Throwable e) {
                 log.warn(e, "Error requesting system memory revoking");
             }
         }, 0, 1, TimeUnit.SECONDS);
     }
 
-    public void requestMemoryRevokingIfNeeded(Collection<SqlTask> sqlTasks)
+    @VisibleForTesting
+    void registerPoolListeners()
+    {
+        memoryPools.forEach(memoryPool -> memoryPool.addListener(memoryPoolListener));
+    }
+
+    private void onMemoryReserved(MemoryPool memoryPool)
```

Comment:
As I mentioned in the other commit, I would make all of these methods synchronized.  This management code is fast enough that there should be no reason for it to be concurrent, which will make the job of future readers and maintainers simpler.